src/java/org/apache/cassandra/db/compaction/CompactionIterator.java

```diff
@@ -175,7 +175,7 @@ public UnfilteredRowIterators.MergeListener getRowMergeListener(DecoratedKey par
 
                 CompactionIterator.this.updateCounterFor(merged);
 
-                if (type != OperationType.COMPACTION || !controller.cfs.indexManager.hasIndexes())
+                if (!preserveOrder())
```

Comment:
This would be extremely confusing to readers. Why does order preservation determine if we should create a row listener?

Create a new method called e.g. `rowProcessingNeeded()`, and call it from `preserveOrder()`  as well as here.

Reply:
done